### PR TITLE
Allow running tests from different suites

### PR DESCRIPTION
**Pull request by aidanhs**
This PR:
 - lets you do `python runner.py --help`
 - moves some potentially printing code into `if __name__ == '__main__'`
 - prints help text as soon as possible, so some warnings will come after it
 - lets you run tests from multiple suites on the same command line (e.g. `python runner.py other.test_sixtyfour_bit_return_value asm2.test_time`) and lists any missing tests - previously it just refused to run anything
 - ensures that if you have 256 test failures it won't come back as 0!

### CODE REVIEW

File: tests/runner.py [hunk 1]
```diff
@@ -968,26 +974,36 @@ def show():
...
+
+  # Run the discovered tests
+  testRunner = unittest.TextTestRunner(verbosity=2)
+  for suite in suites:
+    results = testRunner.run(suite)
+    numFailures += len(results.errors) + len(results.failures)
 
   # Return the number of failures as the process exit code for automating success/failure reporting.
-  exit(numFailures)
+  exitcode = numFailures if numFailures < 255 else 255
```
Comment (kripken) [inline tests/runner.py]: max(numFailures, 255)

File: tests/runner.py [hunk 2]
```diff
@@ -968,26 +974,36 @@ def show():
...
+    print >> sys.stderr, 'No tests at all found!'
     numFailures = 1
-  else:
-    testRunner = unittest.TextTestRunner(verbosity=2)
-    for suite in suites:
-      results = testRunner.run(suite)
-      numFailures += len(results.errors) + len(results.failures)
+  elif len(names) > 0:
+    print 'WARNING: could not find the following tests: ' + ' '.join(names)
+    numfailures = len(names)
```
Comment (kripken) [inline tests/runner.py]: capital `F`.

File: tests/runner.py (hunk 1 above)
Comment (aidanhs) [inline tests/runner.py]: I think you mean `min`!

**Comment by aidanhs** [comment]
Review comments addressed.

File: tests/runner.py (hunk 2 above)
Comment (aidanhs) [inline tests/runner.py]: Impressive spot.

File: tests/runner.py (hunk 1 above)
Comment (kripken) [inline tests/runner.py]: heh, yes :)

**Comment by kripken** [comment]
It doesn't seem to print the incorrect tests for me,

````
$ ./tests/runner.py test_asfd
No tests at all found!
````

Let's also add a test for this stuff? Can do it in `other`.

**Comment by aidanhs** [comment]
I'm not sure what you're expecting to see? I guess it's worth figuring out the expected behaviour before I add a test. My idea was:
 - if you completely fail to choose valid tests, print an error saying it has no chance of running tests (what you see)
 - if only some tests are wrong, print the names of those nonexistent tests

I take it you'd prefer to print nonexistent tests even in the first case?

**Comment by kripken** [comment]
Yes, I think it's useful if I ask for a bunch of tests, and have a typo in one of them, that we report that name.

**Comment by aidanhs** [comment]
It already does that - the *only* situation it doesn't print out test names is if *none* of the tests are valid. E.g.

```
$ ./tests/runner.py test_asfd test_ccall
WARNING: could not find the following tests: default.test_asfd
test_ccall (test_core.default) ... (checking sanity from test runner)
[...]
OK
```

**Comment by kripken** [comment]
Oh ok, got it.

The only thing I'd prefer then is for that to be an error. Otherwise if I missing the warning, it still says "OK" at the bottom, so I could miss that I had a typo in some tests.

**Comment by aidanhs** [comment]
Pushed a new commit for better reporting on multi-suite runs. The problem is that the 'OK' comes from the individual suite, so there needs to be an additional report, like the one I've added. It figures out if there's output you might have missed and, if so, adds a summary report.

Example output:
```
$ python runner.py default.qwerty 
WARNING: could not find the following tests: default.qwerty
$ python runner.py default.qwerty default.test_unaligned
WARNING: could not find the following tests: default.qwerty
test_unaligned (test_core.default) ... (checking sanity from test runner)
INFO:root:(Emscripten: Running sanity checks)
WARNING:root:java does not seem to exist, required for closure compiler, which is optional (define JAVA in /myhome/.emscripten if you want it)
<skipping: LLVM marks the reads of s as fully aligned, making this test invalid> ok

----------------------------------------------------------------------
Ran 1 test in 0.073s

OK
====================

TEST SUMMARY
    Could not find 1 tests
    test_core: 1 ran, 0 errors, 0 failures, 0 skipped
$ python runner.py other.test_sixtyfour_bit_return_value asm2.test_time asm1.test_ccall 
default.qwerty default.test_unaligned
[...]
Ran 1 test in 0.607s

FAILED (errors=1)
====================

TEST SUMMARY
    Could not find 1 tests
    test_core: 3 ran, 1 errors, 0 failures, 0 skipped
    test_other: 1 ran, 1 errors, 0 failures, 0 skipped
```

(the oddity with it saying '0 skipped' is because the emscripten tests do not do a 'real' test skip)

**Comment by aidanhs** [comment]
Note that this removes the special handling for 'No tests at all found!' - it'll just print the warnings (and exit with a non-zero exit code).

**Comment by kripken** [comment]
That's great, thanks!

**Pull request merged**